File: crackingthecodeinterview/linkedlists/kthnodegiven.cpp
#include <iostream>

using namespace std;

typedef struct node{
        int  data;
        node *next;
}node;

typedef struct ll{
    node *head;
}ll;

void
remove_item(node &input)
{
    node *nextnode   = NULL;
    node *currnode   = &input;

    while(currnode->next != NULL){
        nextnode       = currnode->next;
        currnode->data = nextnode->data;
        currnode       = nextnode;
    }

    delete currnode;
}

void
insert_item(ll &list, int num)
{
    node *newnode  = new node;
    newnode->data  = num;
    newnode->next  = list.head;
    list.head      = newnode; 
}

void
clean_ll(ll &list)
{
    node *curr = list.head;
    node *next = NULL;
    while(next != NULL){
        curr =  next;
        next = curr->next;
        delete curr;
    }

    if(curr != NULL)
        delete curr;
}

void
print_ll(ll &list)
{
    node *curr = list.head;

    while(curr->next != NULL){
        cout << curr->data;
        curr = curr->next;
    }
    cout << endl;
}
        

int
main(void)
{
    
    ll list; 
    list.head       = new node;
    list.head->data = 10;
    list.head->next = NULL;
    
    insert_item(list,15);
    insert_item(list,20);
    insert_item(list,9);
    insert_item(list,14);
    insert_item(list,12);

    print_ll(list);
    
    //Geting the node to pass to function
    node *currnode = list.head;
    for(int i = 0; i < 3; i++)
        currnode = currnode->next;

    remove_item(*currnode);

    print_ll(list);

    clean_ll(list);

    return 0;
}

File: crackingthecodeinterview/linkedlists/kthnodegiven.py
class node:
    def __init__(self):
        self.next = None
        self.data = None

    def __str__(self):
        return str(self.data)

    def __eq__(self,other):
        if isinstance(other, type(self)):
            return self.next == other.next and self.data == other.data

class ll:
    def __init__(self):
        self.head = None

    def __str__(self):
        node = self.head        
        data = ""
        while node != None:
            if(str(node) == "None"):
                node = node.next
                continue;
            data += str(node) + " "
            node = node.next
        return data

    def add_node(self,data):
        if self.head == None:
            self.head      = node()
            self.head.data = data
            return
        
        new_node      = node()
        new_node.data = data
        new_node.next = self.head
        self.head     = new_node

    def remove_node(self,data):
        node = self.head
        if(node == None):
            return

        while node.data != data:
            node = node.next 

        if(node == None):
            return

        while node.next != None:
            node.data = node.next.data
            node      = node.next
        
        node.data = None
        node      = None 

def main():

    list = ll()

    list.add_node(5)
    list.add_node(8)
    list.add_node(12)
    list.add_node(15)
    
    print(list)

    list.remove_node(8)

    print(list)

main()
